compiler/rustc_monomorphize/src/partitioning.rs

```diff
@@ -1179,23 +1179,19 @@ fn collect_and_partition_mono_items(tcx: TyCtxt<'_>, (): ()) -> MonoItemPartitio
         }
     }
 
+    let mut autodiff_mono_items: Vec<_> = vec![];
     let mono_items: DefIdSet = items
         .iter()
         .filter_map(|mono_item| match *mono_item {
-            MonoItem::Fn(ref instance) => Some(instance.def_id()),
+            MonoItem::Fn(ref instance) => {
+                autodiff_mono_items.push((mono_item, instance));
```

Comment:
Why this collected unconditionally? What if i don't use autodiff feature?

Look like this thing collects `instance` twice: inside `mono_item` and second one - separately as `instance`.

Reply:
it's a per-function thing so we at least need to iterate over all functions to determine which ones we need to keep. I guess we could do it without allocating a list by just iterating over all mono items and filtering there

Reply:
The intention is to flip autodiff in the next days/weeks anyway to ship it on all nightly builds. The collection here has no measurable perf impact anymore, but if you care I can add a cfg to only push if we build with llvm_enzyme enabled.

Yes, some elements are duplicated, but mono_instances also contains static elements, whereas autodiff does not want them. Also autodiff wants both mono_item and instances, whereas the mono_items vector does not. Filtering thr mono_items in a second iteration might have been the culprit for the perf regression, so this code seems to be the best trade off.

Reply:
> The intention is to flip autodiff in the next days/weeks anyway

Until i use autodiff nightly feature, it's nice to avoid any related code.

>Yes, some elements are duplicated, but mono_instances also contains static elements

No, i'm saying about line `autodiff_mono_items.push((mono_item, instance));`: where `instance` from `MonoItem::Fn(ref instance)`, which is literally `mono_item`, so you pushing `mono_item` and it's `instance` both.

Reply:
It seems like I can't (?) get the mono_item out of an instance, or did I miss the function?
I need the item to call `usage_map.used_map.get(&item)`. I could match against the item inside of the using function to get access to the instance, but I'm already doing that once, so it seems bad to repeat the matching? 

I updated the code to not push. I must create two variables now because the linter isn't smart enough that in one version the vector is mutated. Allowing the unused_mut in such a large file isn't great and expect(unused_mut) also doesn't work because in one config it is used.

Reply:
I think the cost is below the noise, but the change is ok, so let's go with it